platform/engineering/design-docs/2020-05-26-content-validation.md

```diff
@@ -0,0 +1,388 @@
+
+In summary, this approach will have the following impact for the following groups:
+
+- **Content Writers**: Improve error reporting while maintaining current convenience
+- **Frontend Tools team**: Reduce build noise, failure rate, and number of builds requiring manual intervention
+- **VA.gov users**: Maintain existing time to see updated content and risk of seeing accessibility errors
+- **Anybody trying to ship code for the VA.gov front end**: Reduce failed deploys due to accessibility errors which should allow them to ship faster
+
+## Specifics
+
+### Detailed Design
+
+#### Reuse existing broken link check
+
+Since broken links already block deploys during content deploys, we will preserve the functionality.
+
+#### New build script to supplement existing build script
+
+Before axe checks can be run, the content needs to be built. A new build script will be created to supplement the existing build script. The new build script will be derived from the existing build script, and modified to only include the steps necessary for the axe checks. Since the existing build script uses Metalsmith, the new build script will also use Metalsmith.
+
+#### Content to be validated
+
+To preserve the current functionality, we will first try to pull the latest content from Drupal. If we are unable to pull the latest content from Drupal, we will use the cached Drupal content.
+
+#### Update accessibility tests config
+
+The [current accessibility tests](https://github.com/department-of-veterans-affairs/vets-website/blob/f93e0c11d9e037bd2460ede5e2fb4b67fbaaf118/Jenkinsfile#L82-L84) use Nightwatch. The default value for [Nightwatch's `end_session_on_fail` test session setting](https://nightwatchjs.org/gettingstarted/configuration/#test-session-settings) is `true`. Since we want a report of all the failures, we would need to update the `end_session_on_fail` setting in our [Nightwatch config](https://github.com/department-of-veterans-affairs/content-build/blob/master/config/nightwatch.docker-compose.js) from `true` to `false`.
+
+#### Improved invalid content reporting
+
+Currently, accessibility errors need to be manually communicated from the Frontend Tools team to the CMS team. And broken links need to be extracted from the build log. We would improve that reporting by generating a file using Winston, and send that file to the #cms-team channel using existing Slack integrations.
+
+#### New scheduled job in Jenkins
+
+Lastly, we will write a scheduled Jenkins workflow to run the content validation script hourly during workdays. That way, the content team learns about invalid content at a convenient cadence.
```

Comment:
Can we run scheduled jobs in CircleCI? (i.e. what do we do post-Jenkins, since Ops team is tearing that down before too much longer?)

Reply:
CircleCI does support scheduled workflows. 

https://circleci.com/docs/2.0/workflows/#scheduling-a-workflow

I actually attempted a prototype of a CircleCI scheduled job in a [branch](https://github.com/department-of-veterans-affairs/vets-website/compare/add-nightly-build-to-circleci) a couple months ago.

Reply:
great! is there a reason we wouldn't just implement this in CircleCI to start?

Reply:
🤔 

On one hand, I can't cite a concrete reason off the top of my head. 

On the other hand, I could see an argument that switching CI tools is slightly outside the scope/spirit of this design doc, and that task would be a good candidate for the **Future Work** section.

Reply:
my 2c: it seems like a strong likelihood that we're going to switch fully to CircleCI. so options are:
- implement this in Jenkins now, have to move it to CircleCI later
- implement in CircleCI, don't have to move later
if that's the case, should we just implement in Circle right away and save the trouble?

Reply:
That makes sense. I've updated the Detailed Design to specify CircleCI instead of Jenkins in https://github.com/department-of-veterans-affairs/va.gov-team/pull/9658/commits/bdc89d7347e9ca7df9ac23524d52eaf848fafe3e.